src/ripple/protocol/impl/BuildInfo.cpp

```diff
@@ -77,82 +77,88 @@ getFullVersionString()
     return value;
 }
 
+static constexpr std::uint64_t implementationVersionIdentifier =
+    0x183B'0000'0000'0000LLU;
+static constexpr std::uint64_t implementationVersionIdentifierMask =
+    0xFFFF'0000'0000'0000LLU;
+
 std::uint64_t
-getEncodedVersion()
+encodeSoftwareVersion(char const* const versionStr)
 {
-    static std::uint64_t const cookie = []() {
-        std::uint64_t c = 0x183B000000000000;
+    std::uint64_t c = implementationVersionIdentifier;
 
-        beast::SemanticVersion v;
+    beast::SemanticVersion v;
 
-        if (v.parse(versionString))
-        {
-            if (v.majorVersion >= 0 && v.majorVersion <= 255)
-                c |= static_cast<std::uint64_t>(v.majorVersion) << 40;
+    if (v.parse(std::string(versionStr)))
+    {
+        if (v.majorVersion >= 0 && v.majorVersion <= 255)
+            c |= static_cast<std::uint64_t>(v.majorVersion) << 40;
 
-            if (v.minorVersion >= 0 && v.minorVersion <= 255)
-                c |= static_cast<std::uint64_t>(v.minorVersion) << 32;
+        if (v.minorVersion >= 0 && v.minorVersion <= 255)
+            c |= static_cast<std::uint64_t>(v.minorVersion) << 32;
 
-            if (v.patchVersion >= 0 && v.patchVersion <= 255)
-                c |= static_cast<std::uint64_t>(v.patchVersion) << 24;
+        if (v.patchVersion >= 0 && v.patchVersion <= 255)
+            c |= static_cast<std::uint64_t>(v.patchVersion) << 24;
 
-            if (!v.isPreRelease())
-                c |= static_cast<std::uint64_t>(0xC00000);
+        if (!v.isPreRelease())
+            c |= static_cast<std::uint64_t>(0xC00000);
 
-            if (v.isPreRelease())
+        if (v.isPreRelease())
+        {
+            std::uint8_t x = 0;
+
+            for (auto id : v.preReleaseIdentifiers)
             {
-                std::uint8_t x = 0;
+                auto parsePreRelease = [](std::string_view identifier,
+                                          std::string_view prefix,
+                                          std::uint8_t key,
+                                          std::uint8_t lok,
+                                          std::uint8_t hik) -> std::uint8_t {
+                    std::uint8_t ret = 0;
+
+                    if (prefix != identifier.substr(0, prefix.length()))
+                        return 0;
+
+                    if (!beast::lexicalCastChecked(
+                            ret,
+                            std::string(identifier.substr(prefix.length()))))
+                        return 0;
```

Comment:
FWIW, this `return 0` is not exercised by the unit tests.

Reply:
Thanks, added a test case.